core/git_command.py

```diff
@@ -245,38 +254,40 @@ def _repo_path(self, throw_on_stderr=True):
         Return the absolute path to the git repo that contains the file that this
         view interacts with.  Like `file_path`, this can be overridden by setting
         the view's `git_savvy.repo_path` setting.
-        """
-        def invalid_repo():
-            if throw_on_stderr:
-                raise ValueError("Unable to determine Git repo path.")
-            return None
 
+        Do not raise error when throw_on_stderr is `False`. It is needed
+        in GsUpdateStatusBarCommand, otherwise, spurious popup will be shown.
+        """
         # The below condition will be true if run from a WindowCommand and false
         # from a TextCommand.
         view = self.window.active_view() if hasattr(self, "window") else self.view
+        window = view.window()
         repo_path = view.settings().get("git_savvy.repo_path")
 
         if not repo_path or not os.path.exists(repo_path):
-            working_dir = self.find_working_dir()
-            if not working_dir or not os.path.isdir(working_dir):
-                return invalid_repo()
-
-            stdout = self.git(
-                "rev-parse",
-                "--show-toplevel",
-                working_dir=working_dir,
-                throw_on_stderr=throw_on_stderr
-                )
+            file_name = view.file_name()
 
-            repo_path = stdout.strip()
+            # try the current file first
+            if file_name:
+                repo_path = self.find_git_toplevel(
+                    os.path.dirname(file_name), throw_on_stderr=False)
+                # only set `git_savvy.repo_path` for current file
+                if repo_path:
+                    view.settings().set("git_savvy.repo_path", os.path.realpath(repo_path))
 
+            # fallback: use the first folder if the current file is not inside a git repo
             if not repo_path:
-                return invalid_repo()
+                if window:
+                    folders = window.folders()
+                    if folders:
+                        repo_path = self.find_git_toplevel(
+                            folders[0], throw_on_stderr=throw_on_stderr)
```

Comment:
I'm sorry but I don't get this. Both `find_working_dir` and the code here *first* try to find a directory of the open file and only after this fails will they look for an open folder in the window.

The call (`view.settings().set("git_savvy.repo_path", os.path.realpath(repo_path))`) sets the repo path for the **view**, **not for the window**; so, if the `.bashrc` file is currently open the setting applies to the open file buffer only, and if the `.bashrc` is in a git repo (e.g a dotfiles repo), then the git commands should run on this repo. If we close the file or move to another file in the current window, then we should try to find it again. Finally, If the `.bashrc` is moved to another window, I'm not sure what usually happens to the view, whether it's reused or reconstructed, but we should be able to handle both cases with the same logic.

Solely from a DRY code perspective, I cannot see a reason not to refactor these to reuse `find_working_dir ` since they do exactly the same.